Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,517 @@
+        if isInfinity.location != NSNotFound {
+            if sign < 0 {
+                self.value = Self.negativeInfinity
+                return
+            }
+            self.value = Self.infinity
+            return
+        }
+
+        var exponentSign = 1
+        let exponentSignRange = match.range(at: REGroups.exponentSign.rawValue)
+        if exponentSignRange.location != NSNotFound, let range = Range(exponentSignRange, in: data) {
+            exponentSign = String(data[range]) == "-" ? -1 : 1
+        }
+
+        let decimalPartNSRange = match.range(at: REGroups.decimalPart.rawValue)
+        guard decimalPartNSRange.location != NSNotFound,
+            let decimalPartRange = Range(decimalPartNSRange, in: data) else {
+            throw BSONError.InvalidArgumentError(message: "Syntax Error: Missing digits in front of the exponent")
+        }
+        let decimalPart = String(data[decimalPartRange])
+        var digits = try Self.convertToDigitsArray(decimalPart)
+
+        var exponent = 0
+        let exponentPartRange = match.range(at: REGroups.exponentPart.rawValue)
+        if exponentPartRange.location != NSNotFound, let range = Range(exponentPartRange, in: data) {
+            exponent = exponentSign * (Int(data[range]) ?? 0)
+        }
+        if let pointIndex = decimalPart.firstIndex(of: ".") {
+            // move the exponent by the number of digits after the decimal point
+            // so we are looking at an "integer" significand, easier to reason about
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+        }
+
+        while exponent > Self.exponentMax && digits.count <= Self.maxSignificandDigits {
+            // Exponent is too large, try shifting zeros into the coefficient
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < Self.exponentMin && !digits.isEmpty {
+            // Exponent is too small, try taking zeros off the coefficient
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = Self.exponentMin
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error: Value too small")
+            }
+        }
+
+        guard (Self.exponentMin...Self.exponentMax).contains(exponent) else {
+            throw BSONError.InvalidArgumentError(message: "Rounding Error: Cannot round exponent \(exponent) further")
+        }
+
+        guard digits.count <= Self.maxSignificandDigits else {
+            throw BSONError.InvalidArgumentError(
+                message: "Overflow Error: Value cannot exceed \(Self.maxSignificandDigits) digits"
```

Comment:
this message may be confusing because the user might have given us a number with fewer digits than this but the number of digits changed in the exponent manipulation above to be > 34. perhaps we should just have two generic underflow and overflow errors we reuse throughout here?